UndoManager Improvements

aaalsaleh opened the pull request:
This pull request implements #334, and provides a form of improvement for the undo manager behavior as suggested in #335:

1. Replaces the current undo manager with a new undo manager. The new undo manager follows the W3C UndoManager specifications, and is more flexible than the current one. See: https://github.com/aaalsaleh/undo-manager
2. Updates the way transactions are pushed into the new undo manager.
3. Adds an `interval` option for the undo manager. Now, consecutive transactions within `interval` will be merged as a single transaction (#335).
4. ~~Adds a new `setUndoManager` API for replacing the internal undo manager with external undo manager (#334).~~
5. Several other tiny updates over the undo/redo core commands, and core/patch events. These are needed for the new undo manager.

Please note that point 2 and 3 can be further improved and optimized specifically for Scribe. For example, ~~transactions content can be replaced instead of merged to save space~~ (done!). Also, merging condition can be improved to take other factors into account.

aaalsaleh commented:
The build failed in e65b169 due to `Uncaught UnknownError: Sauce Labs Authentication Error`. Otherwise, the commit is correct and exactly like 00e08f2, which didn't fail. There is no way to force rebuild without another commit or closing and reopening this PR...

hmgibson23 commented on src/scribe.js:
```diff
@@ -179,10 +181,19 @@ define([
+   * @param {Object} undoManager The undo manager instance to be used by the current editor.
+   */
+  Scribe.prototype.setUndoManager = function (undoManager) {
+	  this.undoManager = undoManager;
```
We use two spaces for alignment in scribe. You should be able use to editorconfig to sort this out for you...

hmgibson23 commented on src/undo-manager.js:
```diff
@@ -1,43 +1,73 @@
+			else
+				stack.unshift([transaction]);
+			this.position = 0;
 
```
Indentation is quite strange again from line 15 - 22.

cutandpastey commented:
Hey, Thanks for the PR! In general I like the approach there are, however, a few issues:

- All the undo-manager tests in `app/undo-manager.spec.js` fail for me here.
- There are several style issues throughout the code, mainly indentation and brackets around `if` statements.
- Personally I'd like to see the `UndoManager` methods moved onto the prototype rather than being defined within the constructor, but this is opinion. Not sure how @hmgibson23 @rrees and @robinedman feel about this?

If you can do a quick tidy and refactor the tests I'll get this merged ASAP!

hmgibson23 commented on src/undo-manager.js:
```diff
@@ -1,43 +1,73 @@
+				this.length = stack.length = limit;
+			else
+				this.length = stack.length;
 
```
We alway use curly braces in for our if statements

hmgibson23 commented on src/undo-manager.js:
```diff
@@ -1,43 +1,73 @@
-      }
-    };
+			if (fireEvent)
+				undoScopeHost.dispatchEvent(new CustomEvent('DOMTransaction', {detail: {transactions: stack[0].slice()}, bubbles: true, cancelable: false}));
```
Indentation and curly braces

hmgibson23 commented on src/undo-manager.js:
```diff
@@ -1,43 +1,73 @@
+			if (this.position < this.length) {
+				for (var i = stack[this.position].length - 1; i >= 0; i--)
+					stack[this.position][i].undo();
+				this.position++;
```
Can you wrap this for with curly braces. Also the indentation is off

rrees replied:
When we were discussing this internally we thought it was better to use a reversed stack (append at head) rather than a decrementing index loop.

hmgibson23 commented on src/undo-manager.js:
```diff
@@ -1,43 +1,73 @@
-    return UndoManager;
-  };
+				if (fireEvent)
+					undoScopeHost.dispatchEvent(new CustomEvent('undo', {detail: {transactions: stack[this.position - 1].slice()}, bubbles: true, cancelable: false}));
```
Curly braces and indentation.

hmgibson23 commented on src/undo-manager.js:
```diff
@@ -1,43 +1,73 @@
+			this.length = stack.length;
+		};
+	}
+
```
Indentation above.

hmgibson23 commented:
This looks solid to me. It's definitely better than our current solution. I listed a few style gripes but we should probably add a style guide anyway. I'm also with @cutandpastey about moving the methods onto the prototype as it's more idiomatic.

rrees commented on src/scribe.js:
```diff
@@ -179,10 +181,19 @@ define([
+   * Replace the current undo manager. Can be used by undo manager plugins or directly by the user.
+   * @param {Object} undoManager The undo manager instance to be used by the current editor.
+   */
+  Scribe.prototype.setUndoManager = function (undoManager) {
```
I don't think it should be possible to change the UndoManager outside initialisation. Not a blocker to merging but not something we should support either.

rrees commented on src/scribe.js:
```diff
@@ -192,14 +203,28 @@ define([
 
-        this.undoManager.push(html);
+        // Register a new change transaction.
+        var scribe = this;
```
Not a blocker but having bound this to a name I think it would be better to use it on the next line as well

rrees commented:
@aaalsaleh This is amazing! Thanks so much! :shipit:

hmgibson23 commented on src/undo-manager.js:
```diff
@@ -1,43 +1,84 @@
+		this.undo = function () {
+			if (this.position < this.length) {
+				for (var i = stack[this.position].length - 1; i >= 0; i--) {
+					stack[this.position][i].undo();
```
So if you have a reversed append as @rrees suggested, you could just a forEach here.

aaalsaleh commented:
@hmgibson23 Thanks for pointing out these weird indentations. It was caused by my vim config.

@cutandpastey 
* It could be just due to the fact the undo manager was replaced with a different one. Could be also due to the behavior of the helpers used in the tests. Of course, it could be that I didn't handle all the cases properly while updating the code.
* I fixed all the coding style issues. Kindly point out if I missed any of them.
* The main reason I chose the constructor approach, is to have the stack and other options private. I don't think it is possible to have private properties with prototype approach. Nonetheless, I can move everything into prototype by making the private variables public, and just prepending them with `_` if needed.

@rrees I removed `setUndoManager` as noted. I originally intended to add the undo manager as a constructor option, but was not sure this will be accepted. I really didn't get your other remark about using the named `this` in the line below...

cutandpastey commented on src/scribe.js:
```diff
@@ -192,14 +195,28 @@ define([
-        this.undoManager.push(html);
+        // Register a new change transaction.
+        var scribe = this;
+        this.undoManager.transact({
```
As in `scribe.undoManager` rather than `this.undoManager`.

cutandpastey commented:
@aaalsaleh `undoManager.stack` was public before your amends so I don't think you need to worry about that being private and I can't see the value of hiding the check either.

aaalsaleh commented:
@cutandpastey Yes, I know. But it is not only hiding the check, it is about hiding the whole stack from direct modifications. They were moved to prototype earlier.

I am still unfamiliar with the details of how scribe works internally, and therefore, I would suggest extensively testing all the special cases. I removed some lines from core/events and patch/events, that were causing issues with the new undo manager. I am not quite sure what special cases they were handling, but without them the new undo manager works as expected. Please make sure you don't merge something adding such feature, only to break the main feature of the editor later on!

rrees commented:
@aaalsaleh I've copied your branch to my local machine and it looks like there is just one failing test now where the undo manager restores the uncleaned html (with style attributes) rather than the cleaned.

aaalsaleh commented:
@cutandpastey Turned out the tests in `test/undo-manager.spec.js` are indeed failing because their expectations are not the same after the change to the new undo manager. The default interval option for merging consecutive transactions is 1000 (1 second), and it appears all of them perform editing faster than this interval, then excepts undo to unapply a single change (which is not a correct expectation). For now, I set the default interval to be 0, and this will not merge any transaction, and will yield the expected result. I think maybe there is a need to take interval into consideration in future tests.

However, a new test has failed, and only when interval is 0. When set to 1000, that test didn't fail, but the previous 3 did. I think it has to do with some sort of event timing, and I'll look into it later on.

aaalsaleh commented:
@rrees It seems that if you change the intervals, then you pass/fail some tests. For example, once I set the default interval to 50, all the tests passed on my machine (last commit). Also, I tried to set the interval to 1000, and then test all the failing test manually by hand, and they all produced the expected output. Of course running it as an automated tests did not produce the expected output. I think there must be a race condition somewhere, and maybe it is related to one of the few setTimout callbacks?

rrees commented:
@aaalsaleh we'll need to get the tests passing reliably before we can merge the change. Feel free to change the tests to match the new behaviour but talking about a race condition makes me feel very nervous.

aaalsaleh commented:
@rrees @cutandpastey @hmgibson23 I've made major changes around the transaction manager, and now all the tests are passing.

The problems were:

1. `core/events` and `patch/events` had some issues differentiating between `setContent` and other mutations. Modifying all of them, and adding `forceMerge` option to `transactionManager.run` solved it.
2. `test/undo-manager.spec.js` was failing due to the assumption that two consecutive transactions should be considered two separate transactions. This is not the case with the new undo manager, especially with the default interval of 1000ms. Either change the tests to add some delay between the two typing events (> 1000ms), or what I did, change the interval to be 0 (no merge). This way, the two consecutive transactions will be considered two separate transactions.

I've also improved `pushHistory` by:

1. Saving more memory space by replacing last transaction's content when merging (instead of just pushing them together).
2. Checking now take into account the possibility of a shared undo manager. If the previous transaction was not from the same instance, or even was not a scribe transaction, the new transaction will not be considered a new one (not merged).

rrees commented:
Great, thanks. I'll check it out locally as soon as I can

aaalsaleh commented:
I've had to make some additional tweaks to handle edge cases. I've also add a new option `manager` under `undo` option. It can be used to pass a user defined undo manager. This will allow sharing a single undo manager instantiated by the user with multiple scribe instances (or other components), if needed. Of course, the shared undo manager must use the same undo manager API, or just reuse the one with scribe. I've quickly tested sharing a single undo manager with two scribe instances, and I can confirm it works as expected. Please let me know if you have a better alternative.

aaalsaleh commented:
More space/time efficiency improvements. **I'll no longer do any more commits, unless of course someone point a problem**. I've tested it with the automated tests with both Chrome and Firefox, and all tests were passed. One exception, is that `undo-manager.spec.js` with Firefox cannot accept options for scribe (Object Freezing as mentioned in the helpers source code). This is not the case with Chrome. Passing an the option `interval` of 0 is important for that test as I mentioned in another comment.

I've also manually tested more undo manager cases, and even put together a test example where two scribe instances share the same undo manager. There is a simple visualization of the undo stack in that example. Please check the `examples/shared-undo-manager.html`.

rrees commented:
@aaalsaleh You're right we should probably be merging rather than using defaults as regarding options. You're solution makes sense given what we do know.

I'm going to try and merge this in.

aaalsaleh commented:
@rrees Very glad to contribute something useful for such great project. Thanks for the opportunity.

rrees commented:
Sadly this didn't pass the tests for FF31 and 32 so we need to check how much we need to support those browsers

aaalsaleh commented:
Can you please list these tests, maybe it is something trivial to fix?

rrees commented:
Here's the [link to the failing build](https://travis-ci.org/guardian/scribe/builds/53633191). We are in the process of depreciating support for 31 and 32 so if you can fix it, great. Otherwise I think we will take this change once we've notified users of the change.

aaalsaleh commented:
If they are "950.1" and "950.2" in the link you provided, then it is just an issue with your helpers with Firefox. Passing the `options` parameter to `helpers.initializeScribe.bind` always freeze the test in Firefox. I've mentioned this with my last commit. It is a problem with the helpers, not the tests files or my commit. However, if it is another problem, I didn't see it in the link.

If you are willing to help me by modify the `scribe-test-harness/helpers.js` to add an additional helper function, then I fix the Firefox issue. Please add the following to that file:
```javascript
exports.setUndoInterval = function (interval) {
  return exports.driver.executeScript(function (interval) {
    window.scribe.options.undo.interval = interval;
  }, interval);
}
```
Once you do, I'll commit a change to `test/undo-manager.spec.js`, which will use this helper method to change the undo interval to be 0 before starting any of the undo manager tests.

I've tested this locally with firefox 31, and I can confirm it is passing with this fix. I can make a commit easily to `test/undo-manager.spec.js`, but not to `scribe-test-harness/helpers.js` since it is only a dev dependency.

rrees commented:
Hi there, I don't quite understand the issue. Ultimately we don't want options to changed after initialisation so I wouldn't want to add this but instead fix the initialisation issue.

cutandpastey commented:
@aaalsaleh you can do the above in the actual test without adding to the test harness.

aaalsaleh commented:
@cutandpastey haha, yes indeed, how did I miss that! I've just did that successfully.
@rrees no need, @cutandpastey was right, I already made the required change. Should I issue a new PR, or maybe reopen this one?

rrees commented:
Maybe submit a new PR and reference this one. Thanks!
